src/odbc/src/app/application_data_buffer.cpp

```diff
@@ -260,10 +260,8 @@ ConversionResult::Type ApplicationDataBuffer::PutStrToStrBuffer(
   if (ANSI_STRING_ONLY) {
     bytesRequired = value.length() * outCharSize;
   } else {
-    thread_local std::wstring_convert<std::codecvt_utf8<wchar_t >, wchar_t>
-      converter;
-    std::wstring inString = converter.from_bytes(value.c_str());
-    bytesRequired = inString.length() * outCharSize;
+    size_t maxExpansionFactor = 4; // Safe upper bound (utf-8 to utf-32)
+    bytesRequired = (value.length() * maxExpansionFactor + 1) * outCharSize;
```

Comment:
```suggestion
    bytesRequired = (std::mbstowcs(nullptr, value.c_str(), 0) * outCharSize);
```

Rather than using the maximum upper bound, we can use `mbstowcs` without a target buffer to avoid the initial conversion, and avoid the wide string copy all together.

Reply:
`mbstowcs` is definitely better than `wstring_convert`, but is still performing a conversion and thus less performant than using a max upper bound (trading some minor space complexity). I'm good with either approach, could even run some benchmarks to compare if you're interested.

Reply:
From [cppreference](https://en.cppreference.com/w/c/string/multibyte/mbstowcs): `if dst is a null pointer, this function returns the number of wide characters that would be written to dst, if converted`. 

Also from [microsoft cpp reference](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/mbstowcs-mbstowcs-l?view=msvc-170): `If the wcstr argument is NULL, mbstowcs returns the number of wide characters that would result from conversion, not including a null terminator`.

If the implementation is compliant with the standard, I think it should not complete the conversion, though without looking at the exact ISO reference, there could be room for error.

We can either calculate the bytes required to ensure we don't convert, or use `mbstowcs`. Using the maximum upper bounds does not align with the [ODBC](https://learn.microsoft.com/en-us/sql/odbc/reference/syntax/sqlgetdata-function?view=sql-server-ver16#retrieving-data-with-sqlgetdata) specification for SQLGetData.